src/CommonLib/hostfxr_utility.cpp

```diff
@@ -54,160 +24,160 @@ HOSTFXR_UTILITY::FindHostFxrDll(
+    STRU                        struHostFxrPath;
     STRU                        strDotnetExeLocation;
     STRU                        strHostFxrSearchExpression;
     STRU                        strHighestDotnetVersion;
-    PWSTR                       pwzDelimeterContext = NULL;
-    PCWSTR                      pszDotnetLocation = NULL;
-    PCWSTR                      pszDotnetExeString(L"dotnet.exe");
-    DWORD                       dwCopyLength;
-    BOOL                        fFound = FALSE;
     HANDLE                      hFileHandle = INVALID_HANDLE_VALUE;
-    SECURITY_ATTRIBUTES         saAttr;
     std::vector<std::wstring>   vVersionFolders;
+    DWORD                       dwPosition;
+    DWORD                       dwLength;
+    WCHAR                       pszDotnetLocation[MAX_PATH];
 
-    if (FAILED(hr))
+    if ((hFileHandle = UTILITY::CheckIfFileExists(pConfig->QueryProcessPath())) != INVALID_HANDLE_VALUE)
     {
-        goto Finished;
-    }
-
-    // Get the System PATH value.
-    if (!UTILITY::GetSystemPathVariable(L"PATH", &struSystemPathVariable))
-    {
-        hr = ERROR_BAD_ENVIRONMENT;
-        goto Finished;
-    }
-
-    // Split on ';', checking to see if dotnet.exe exists in any folders.
-    pszDotnetLocation = wcstok_s(struSystemPathVariable.QueryStr(), L";", &pwzDelimeterContext);
-    while (pszDotnetLocation != NULL)
-    {
-        dwCopyLength = (DWORD) wcsnlen_s(pszDotnetLocation, 260);
-
-        // We store both the exe and folder locations as we eventually need to check inside of host\\fxr
-        // which doesn't need the dotnet.exe portion of the string
-        hr = strDotnetExeLocation.Copy(pszDotnetLocation, dwCopyLength);
+        hr = UTILITY::ConvertPathToFullPath(L"hostfxr.dll", pConfig->QueryApplicationPath()->QueryStr(), &struHostFxrPath);
         if (FAILED(hr))
         {
             goto Finished;
         }
 
-        if (dwCopyLength > 0 && pszDotnetLocation[dwCopyLength - 1] != L'\\')
+        if ((hFileHandle = UTILITY::CheckIfFileExists(&struHostFxrPath)) != INVALID_HANDLE_VALUE)
         {
-            hr = strDotnetExeLocation.Append(L"\\");
-            if (FAILED(hr))
+            // Standalone application
+            if (FAILED(hr = pConfig->QueryHostFxrLocation()->Copy(struHostFxrPath)))
             {
                 goto Finished;
             }
-        }
 
-        hr = struHostfxrPath->Copy(strDotnetExeLocation);
-        if (FAILED(hr))
-        {
+            hr = GetStandaloneHostfxrParameters(pConfig);
             goto Finished;
         }
-
-        hr = strDotnetExeLocation.Append(pszDotnetExeString);
-        if (FAILED(hr))
+        else
         {
-            goto Finished;
+            UTILITY::ConvertPathToFullPath(pConfig->QueryProcessPath()->QueryStr(), pConfig->QueryApplicationPath()->QueryStr(), &strDotnetExeLocation);
         }
+    }
+    else if ((dwLength = SearchPath(NULL, L"dotnet", L".exe", MAX_PATH, pszDotnetLocation, NULL)) == 0)
```

Comment:
please handle small buffer size issue and do retry. Otherwise you may block long path scenario